src/Model/Item.php

```diff
@@ -1497,88 +1480,50 @@ private static function getDuplicateID(array $item)
 	}
 
 	/**
-	 * Fetch parent data for the given item array
+	 * Fetch top-level parent data for the given item array
 	 *
 	 * @param array $item
 	 * @return array item array with parent data
+	 * @throws \Exception
 	 */
-	private static function getParentData(array $item)
+	private static function getTopLevelParent(array $item)
 	{
-		// find the parent and snarf the item id and ACLs
-		// and anything else we need to inherit
-
-		$fields = ['uri', 'parent-uri', 'id', 'deleted',
+		$fields = ['uid', 'uri', 'parent-uri', 'id', 'deleted',
 			'allow_cid', 'allow_gid', 'deny_cid', 'deny_gid',
 			'wall', 'private', 'forum_mode', 'origin', 'author-id'];
-		$condition = ['uri' => $item['parent-uri'], 'uid' => $item['uid']];
+		$condition = ['uri' => $item['thr-parent'], 'uid' => $item['uid']];
 		$params = ['order' => ['id' => false]];
 		$parent = self::selectFirst($fields, $condition, $params);
 
 		if (!DBA::isResult($parent)) {
-			Logger::info('item parent was not found - ignoring item', ['parent-uri' => $item['parent-uri'], 'uid' => $item['uid']]);
+			Logger::notice('item parent was not found - ignoring item', ['thr-parent' => $item['thr-parent'], 'uid' => $item['uid']]);
 			return [];
-		} else {
-			// is the new message multi-level threaded?
-			// even though we don't support it now, preserve the info
-			// and re-attach to the conversation parent.
-			if ($parent['uri'] != $parent['parent-uri']) {
-				$item['parent-uri'] = $parent['parent-uri'];
-
-				$condition = ['uri' => $item['parent-uri'],
-					'parent-uri' => $item['parent-uri'],
-					'uid' => $item['uid']];
-				$params = ['order' => ['id' => false]];
-				$toplevel_parent = self::selectFirst($fields, $condition, $params);
-
-				if (DBA::isResult($toplevel_parent)) {
-					$parent = $toplevel_parent;
-				}
-			}
-
-			$item['parent']        = $parent['id'];
-			$item["deleted"]       = $parent['deleted'];
-			$item["allow_cid"]     = $parent['allow_cid'];
-			$item['allow_gid']     = $parent['allow_gid'];
-			$item['deny_cid']      = $parent['deny_cid'];
-			$item['deny_gid']      = $parent['deny_gid'];
-			$item['parent_origin'] = $parent['origin'];
+		}
 
-			// Don't federate received participation messages
-			if ($item['verb'] != Activity::FOLLOW) {
-				$item['wall'] = $parent['wall'];
-			} else {
-				$item['wall'] = false;
-			}
+		if ($parent['uri'] == $parent['parent-uri']) {
```

Comment:
When this is the case, you currently aren't checking for `isAllowedByUser`.

Reply:
Yes, in `isValid` above.

Reply:
But that only checks for the user of this parent. This doesn't work if there is a local post from another user with this parent-uri. This checks needs to be applied here as well: https://github.com/friendica/friendica/pull/9515/files#diff-785d24b46fdf749363656f33208f941ea0eca809c7c76452180a0bfc16df68a3R1519-R1522

Reply:
Aaaah, I see what you mean now. So I need to perform this check outside of the `getTopLevelParentData` method.

Reply:
Done.